### PR TITLE
Support playing remote media

### LINKED ISSUE #1559: Cannot play remote videos via QtMultimedia
- Device: BQ 4.5, BQ M10 HD
- Channel: stable
- Build: OTA-12

### Steps to reproduce
1. Copy the attached `test.qml` file into the device
2. from a terminal (or SSH), run `qmlscene ./test.qml --desktop_file_hint=/usr/share/applications/morph-browser.desktop`

### Expected behavior
Video plays.

### Actual behavior
Video does not play.

### Logfiles and additional information
This is a video taken with the BQ 4.5's camera. If the video is copied locally, and the source path is changed to `file:///home/phablet/video20200912_002904356.mp4`, then the video plays.

But also, none of the videos from https://file-examples.com/index.php/sample-video-files/ seem to be playable.

The same program works flawlessly on a desktop installation of Qt 5.9.
[Terminal output](https://github.com/ubports/ubuntu-touch/files/5212346/app.log)
[media-hub.log](https://github.com/ubports/ubuntu-touch/files/5212477/media-hub.log)

[test.qml.gz](https://github.com/ubports/ubuntu-touch/files/5212347/test.qml.gz)

### PR DESCRIPTION
**Pull request by mardy**
This and https://github.com/ubports/qtubuntu-media/pull/5 will fix https://github.com/ubports/ubuntu-touch/issues/1559

### CODE REVIEW

File: src/core/media/player_skeleton.cpp [hunk 1]
```diff
@@ -347,9 +353,31 @@ struct media::PlayerSkeleton::Private
...
+                if (value >= 100 ||
+                    time_since_last_emission > std::chrono::milliseconds(500)) {
+                    // Send the signal immediately
+                    buffering_signal_timeout.cancel();
+                    buffering_signal_value = -1;
+                    remote_buffering_changed->emit(value);
+                    buffering_signal_last_emission = now;
+                } else {
+                    if (buffering_signal_value < 0) {
+                        buffering_signal_timeout.expires_from_now(boost::posix_time::milliseconds(500));
```
Comment (peat-psuwit) [inline src/core/media/player_skeleton.cpp]: Maybe also consider `time_since_last_emission`? If a buffering change comes 400ms after the last emission, I think I want to receive that change 100ms later, not 500ms later.

File: src/core/media/player_skeleton.cpp [hunk 2]
```diff
@@ -347,9 +353,31 @@ struct media::PlayerSkeleton::Private
...
+                auto time_since_last_emission = now - buffering_signal_last_emission;
+                if (value >= 100 ||
+                    time_since_last_emission > std::chrono::milliseconds(500)) {
+                    // Send the signal immediately
+                    buffering_signal_timeout.cancel();
+                    buffering_signal_value = -1;
+                    remote_buffering_changed->emit(value);
+                    buffering_signal_last_emission = now;
+                } else {
+                    if (buffering_signal_value < 0) {
```
Comment (peat-psuwit) [inline src/core/media/player_skeleton.cpp:369]: The buffering will not go below 0, right?

Comment (mardy) [reply src/core/media/player_skeleton.cpp:369]: Correct. But I use `-1` as a sentinel value to indicate that there is no queued signal emission.

File: src/core/media/player_skeleton.cpp (hunk 1 above)
Comment (mardy) [reply src/core/media/player_skeleton.cpp]: Good idea, I'll change it!

Comment (mardy) [reply src/core/media/player_skeleton.cpp]: Done

Comment (peat-psuwit) [reply src/core/media/player_skeleton.cpp]: I did not see the change. Are you sure you pushed to the correct branch?

Comment (mardy) [reply src/core/media/player_skeleton.cpp]: It's more likely that I just forgot to push :-) Done now :-)

**Comment by dobey** [comment]
Would be nice to have test coverage for these changes too. Is there any reasonable way to get some in the gstreamer engine unit tests?

**Comment by mardy** [comment]
> Would be nice to have test coverage for these changes too. Is there any reasonable way to get some in the gstreamer engine unit tests?

To be honest, I find this code to be very difficult to work with. The GStreamerEngine test only tests the `gstreamer::Engine` class, so I would need to write a new test class from scratch for this. Can we postpone it till the time when we rewrite this in Qt+Glib? :-)

**Comment by dobey** [comment]
OK. I'm not familiar enough with the code, so wasn't totally clear how difficult it would be to add tests for this. I don't think Qt would make it easier though. :)

**Review by dobey — APPROVED** [review]

**Pull request merged**